More intuitive relation reference widget text filter

signedav opened the pull request:
It has been quite irritating that with the delete button only the filter text has been emptied, and the value has been still unchanged. 
![bugwithrelationwidget](https://user-images.githubusercontent.com/28384354/46475049-b9e40700-c7e4-11e8-805d-1e14971b4aa0.gif)

I think it's much more intuitive to be able to set the value to NULL (if allowed) by pressing the delete-button. By deleting with backspace or Ctrl+A and Del it's still like before that only the filter-text is edited.

Additionally on editing the filter text the original value always popped in (on every filter-update). See in GIF above Scarlett Johanson comes all the time. This has been quite annoying.

![bugwithrelsolved](https://user-images.githubusercontent.com/28384354/46475199-2ced7d80-c7e5-11e8-9b02-6948c96f5bfe.gif)

signedav committed:
setting of reload interval in featurelistcombo

signedav committed:
set value to null when pressed lineedit-clearbutton

signedav committed:
do not overwrite text entry on editing and model reloadings

signedav committed:
remove setReloadTime because it's not used anymore

DelazJ commented:
@signedav what's that database you and Scarlett Johanson are in together? 🤔

elpaso commented on src/gui/qgsfeaturelistcombobox.cpp:
```diff
@@ -53,6 +53,7 @@ QgsFeatureListComboBox::QgsFeatureListComboBox( QWidget *parent )
 
   mLineEdit = new QgsFilterLineEdit();
   mLineEdit->setSelectOnFocus( true );
+  connect( mLineEdit, &QgsFilterLineEdit::cleared, this, &QgsFeatureListComboBox::setValueToNull );
```
I tend to prefer a lambda in such a simple case, when the slot is only a few lines and it is just used once as a slot, but it's really up to you.

signedav replied:
I agree. When I made it like this, I wasn't sure yet that it's only used once.

elpaso commented:
Looks good to me, thanks!

Just as a side (unrelated) note, I find annoying that you need to delete the `NULL` string before you start typing, I wonder if it could be automatically removed when the user focuses in the line edit (and restored in case the user does not type anything), perhaps you have a good idea about to solve this issue.

signedav committed:
use lambda to set value to null

signedav commented:
Thanks @elpaso - I'll ask Scarlett to join the next hackfest. I'm sure she likes it :slightly_smiling_face: 

Regarding the manual delete of "NULL" you are right. But this is the use case for line edits in general. When you enter the field by tab - then the text is selected. When you enter by mouse click, you are on the position of the mouse. This makes sense, when you want to change "Scarlett Johanson" to "Scarlett Johansson" or "Scarlett Jackson", but does not with null. So we could override the mouse click in case of "NULL" that it:
- selects all text
- removes all text (and in normal text edit empty text will be represented with "NULL" when nothing typed in, and in the `QgsFeatureListCombo` it will take the old value - means "NULL" again)

elpaso commented:
Looking forward to the next HF then :)

Regarding your proposal for NULL overwriting, it looks good, the only problem I see is that we need to distinguish NULL from empty string, and we cannot know after the user focuses on the field if she wanted to change the value from NULL to empty or just focused on the line edit and then changed her mind.

But it's not an issue with the current implementation: I think that the user can always click on the delete-button and change the value back from empty to NULL.

elpaso commented:
> So probably the improvement could be to have the NULL text selected on mouse click instead set cursor to the clicked position.

+1 

> Should I add it to this PR or should we finish this one, since those are not directly related to each other.

Your choice.

signedav committed:
selectAll if requested on mouse release

because otherwise the mouse event makes the selectAll useless if it's done on focusEvent

signedav commented:
I decided to do it here.

Actually I found out, that this is already implemented, but not working. 

On the `focusEvent` it used to call a `selectAll()` - but it de-selected the text after that because of the cursor-placing. So the `selectAll()` has to be called *after* the mouse has been released and that's what is done now.

Btw. on the combo box it selects the text at every value (not just NULL) and otherwise on `QgsFilterLineEdit` it selects if it's NULL:
![selectall](https://user-images.githubusercontent.com/28384354/46723521-58a2b480-cc78-11e8-91f8-14137f848285.gif)

signedav commented:
@elpaso do you approve my last commits?

elpaso commented:
@signedav looks good to me, but I did not have time test them. 
That widget is a bit fragile, would you like me to test them?

andreasneumann commented:
I agree - testing this widget a lot would make sense. Please do testing, @elpaso 

It is a very powerful widget, but also fragile. With every new QGIS release I am afraid that this widget might be broken again ...

elpaso commented:
ok, I'll do that ASAP (in time for 3.4)

signedav commented:
Thanks @elpaso

elpaso commented:
The line edit looks good, can you add this test case?

```python
    def test_selectedText(self):
        """ test that NULL value is selected on focus and not-null value is not"""

        w = qgis.gui.QgsFilterLineEdit(nullValue='my_null_value')
        w.clearValue()
        self.assertEqual(w.selectedText(), 'my_null_value')

        w.setValue('my new value')
        self.assertEqual(w.selectedText(), '')

        w.clearValue()
        self.assertEqual(w.selectedText(), 'my_null_value')

```

Here are the issues I've found in the combo:
- the clear value action indicator (`x`) should not appear when the value is null
- the null value should be selected when the widget gets the focus (just like in the line edit) 
- the above conditions should be added to the test cases in `TestQgsFeatureListComboBox`, I suspect that when calling `cb->setAllowNull( true );` `cb->lineEdit()->setNullValue(...)`  (object_cast may be necessary) should also be called.

signedav commented:
Thanks.

> the null value should be selected when the widget gets the focus (just like in the line edit)

It does select all on getting focus by tap and on getting focus by mouse click, correct?
I have to check if it's selected on `setFocus()` as well...

elpaso commented:
> Thanks.
> 
> > the null value should be selected when the widget gets the focus (just like in the line edit)
> 
> It does select all on getting focus by tap and on getting focus by mouse click, correct?
> I have to check if it's selected on `setFocus()` as well...

I did test with a single widget only (in a dialog) so I could only test focus with `QWidget::setFocus(Qt::FocusReason reason)` and apparently the NULL value is not automatically selected when the widget get the focus (I tested with TabFocusReason too), even if `allowNull(true)` was called, like I mentioned in my previous comments, I suspect that there should be some code in the combo that setups the line edit to handle null correctly.

signedav committed:
pass nullRepresentation to QgsFilterLineEdit

signedav committed:
remove connection to lineeditcleared, because of already correct null value implementation there

signedav committed:
tests for switching between values

especially setting the NULL value and checking if it's selected
checking if the clear button is activ when it should and inactive when it should

signedav commented:
Regarding the tests. This (not implemented) does not work:
```
  QTest::keyClicks( cb.get(), QStringLiteral( "sleeve" ) );
  QTest::keyClick(cb.get(), Qt::Key_Enter );
  waitForLoaded( cb.get() );
  QCOMPARE( cb->model()->data( cb->currentModelIndex() ).toString(), QStringLiteral( "sleeve" )  );
```
Neither it does in QGIS. But does it need to? The index seemed saved on switching between items (attributelist-form) or pressing "okay". But it does not save neither when I enter a text on a TextEdit-Widget and press enter...

signedav committed:
controll clear button by edit-state

because it has to appear on edit-mode, otherwise not

signedav commented:
@elpaso Do you agree on the last changes? 

I passed the `nullRepresentation()` in the constructor, so I guess it's not needed to set it again. And if NULL is not allowed it does not mind to have it defined. With this the index is set after the `cleared` to `nullIndex()`.

I found another issue, that the "clear button" is not displayed when switched to non edit mode and back. 

Regarding the tests is the last proposed part from you missing (selecting another Item from the dropdown). 
But manual tests over QGIS have been successful there.

elpaso commented on src/gui/qgsfeaturelistcombobox.cpp:
```diff
@@ -203,6 +203,14 @@ void QgsFeatureListComboBox::keyPressEvent( QKeyEvent *event )
+  if ( event->type() == QEvent::EnabledChange )
+    mLineEdit->setShowClearButton( true );
+
+  return QComboBox::event( event );;
```
extra `;`

elpaso commented on src/gui/qgsfeaturelistcombobox.cpp:
```diff
@@ -203,6 +203,14 @@ void QgsFeatureListComboBox::keyPressEvent( QKeyEvent *event )
+bool QgsFeatureListComboBox::event( QEvent *event )
+{
+  if ( event->type() == QEvent::EnabledChange )
+    mLineEdit->setShowClearButton( true );
```
I admit I don't follow this: shouldn't you check if the widget is enabled and setShowClearButton accordingly?

elpaso replied:
in any event, make sure that this does not show the clear button when it's not needed (value is NULL).

signedav replied:
As far I understood the `setShowClearButton( true )` it only sets that the clear button is required on the usual circumstances (not null, enabled, not read-only) handled here https://github.com/qgis/QGIS/blob/master/src/gui/qgsfilterlineedit.cpp#L183

It could call `updateClearIcon` here as well but then I need to change it to public.

elpaso replied:
what about implementing this logic directly inside the lineedit? Wouldn't be more logical? Or is there a reason why the `EnabledChange` event does not apply to the lineedit.

elpaso commented:
Can you please also add the python test I pasted above (for line edit) to `test_qgsfilterlineedit.py`?

signedav committed:
Test if NULL is selected

signedav committed:
removed extra ;

signedav committed:
handle EnableChange event in QgsFilterLineEdit

3nids commented:
issue https://issues.qgis.org/issues/20153 is related
tnis would be a must fix!

luipir commented:
> issue https://issues.qgis.org/issues/20153 is related
> tnis would be a must fix!

fully agree, please add test to avoid this kind of regressions

m-kuhn commented:
I would love to merge this, but I'm hesitant to merge this so close to release without in depth testing.

@luipir do you have the possibility to test this?

luipir commented:
> @luipir do you have the possibility to test this?
cherry-piking...

luipir commented:
tested. and related to  https://issues.qgis.org/issues/20153  this PR does not fix it (I didn't expected becasue the focus of the PR is on combobox)

m-kuhn commented:
Thanks @luipir 
Did you have the chance to test the combobox also or did you just check 20153?

luipir commented:
sorry I need to replicate the layer configuration (lack of time) => I checked only 20153! and now I've to finish a not related to qgis work... @signedav may you share at least a test project with which we can spread testing?

m-kuhn commented:
@luipir no worries, I think there are enough other people out there with test data/projects.

I feel confident about it, but it should at least get a couple of nightly builds to test before it's released. I think we'll defer this one to right after 3.4.0 has been released and wait for test results from those nightlies so it's available in 3.4.1

m-kuhn commented:
unless of course @elpaso feels confident about merging this!

signedav commented:
Sorry, for not being available this morning. I think it's working fine, and anyway better than before. But I can life with 3.4.1 as well. Thank you all.

elpaso commented:
Sorry, just got back in the office, I believe this PR is a very good enhancement! But I'd really like some more testing on it, as @luipir  suggested it would be really useful if you could share a test project with different widget configurations to facilitate testing from different devs.

m-kuhn commented:
I'd propose to merge it now. It has been sitting around for almost a month and the feedback from testers was positive. Once it's in the next nightly builds, testing will increase considerably.

elpaso commented:
@m-kuhn +1 for merge, let's give this a field trial and I'm pretty sure we will fix corner cases in time for 3.6

m-kuhn commented:
Thanks for the feedback @elpaso and let's look that we have fixed corner cases by 3.4.1 already ;)

m-kuhn commented:
@signedav can you cherry-pick this to release-3_4 and create a pull request on that branch too?

gacarrillor commented:
@m-kuhn and @signedav Works great! :+1:

signedav commented:
Thanks @gacarrillor
